Fix race condition in TP agent

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59377 Move CUDA-related stuff of TP agent to separate file
* #59376 Make CUDA serde support for TP agent pluggable
* #59375 Merge TensorPipe's CPU and CUDA channel registry
* #59299 Remove LazyStreamContext (2 out of 2)
* #59298 Remove LazyStreamContext (1 out of 2)
* #59297 Fix CUDA sync when switching streams in RPC tests
* **#58753 Fix race condition in TP agent**
* #56863 Ensure async_execution works with CUDAFuture
* #57355 Avoid re-doing CUDA stream sync in OwnerRRef

TSAN was (rightfully!) detecting and complaining about a race due to the fact that upon init the TP agent exchanges the device maps between nodes using RPC requests (and by doing so it accesses the device maps) and then sets the reverse device maps (thus possibly modifying the set of devices). This resulted in a data race, i.e., simultaneously reading and writing the set of devices without synchronizing.

One solution is to add a mutex around the devices, which works, but is "annoying". An alternative solution is to make the set of devices immutable (i.e., `const`). For that to work, we need to exchange the device maps without using RPC calls. We can do so using the process group that we need to create anyways.

Since now there's a lot more logic in Python, I've moved (and restructured) all safety checks over there, and removed them from C++.

Differential Revision: [D28603754](https://our.internmc.facebook.com/intern/diff/D28603754/)

lw committed:
Fix race condition in TP agent

TSAN was (rightfully!) detecting and complaining about a race due to the fact that upon init the TP agent exchanges the device maps between nodes using RPC requests (and by doing so it accesses the device maps) and then sets the reverse device maps (thus possibly modifying the set of devices). This resulted in a data race, i.e., simultaneously reading and writing the set of devices without synchronizing.

One solution is to add a mutex around the devices, which works, but is "annoying". An alternative solution is to make the set of devices immutable (i.e., `const`). For that to work, we need to exchange the device maps without using RPC calls. We can do so using the process group that we need to create anyways.

Since now there's a lot more logic in Python, I've moved (and restructured) all safety checks over there, and removed them from C++.

Differential Revision: [D28603754](https://our.internmc.facebook.com/intern/diff/D28603754/)

[ghstack-poisoned]

lw commented on torch/distributed/distributed_c10d.py:
```diff
@@ -295,9 +295,7 @@ def _get_group_size(group):
-    if group not in _pg_group_ranks:
-        raise RuntimeError("The given group does not exist")
-    return len(_pg_group_ranks[group])
+    return group.size()
```
This change (and the one below) were failing because `all_gather_object` was failing when used with a ProcessGroup instance which was created "by hand", because it didn't have an entry in the global variables `_pg_map` and `_pg_group_ranks`. I don't think there's any fundamental reason for that restriction, hence here I attempted to fix it. Not sure if this is the best way to go about it? CC @rohan-varma

rohan-varma replied:
I think this is fine for now, although ideally we shouldn't have to modify these private APIs to get things to work properly as this is basically a process group that c10d is not "aware of" in some sense.

On the other hand, it looks like rpc's `_init_process_group` needs to do this because we don't want to initialize a default group. One alternate solution here is to register that pg in the private maps, although that's not great either. We could consider just initializing the default group in that case, though that comes with tradeoffs as well.

lw replied:
What I don't understand is why a manually-created process group is not self-sufficient? Why won't it fully work unless some additional metadata is store _somewhere else_? If this metadata is always needed, shouldn't we store it on the ProcessGroup object itself? If it's never needed, should we drop it? If it's only sometimes needed, we should figure out where that's strictly the case, and only make those operations fail, but all other operations should work.

rohan-varma replied:
@lw I think this is a relic from the original design of c10d python package, where a python-side process group is not considered properly registered unless it is created with `init_process_group` or `new_group` APIs. i.e. there is no guaranteed support of process groups created directly from the pybind APIs. Torchrec team ran into a bunch of issues around this internally earlier as well. 

Why this is the case will probably require digging into the original design of c10d python package, but agreed that it would be nice to support this. As a first step we can probably try to document all APIs that break if the process group is manually created using the pybind APIs.

lw committed:
Update on "Fix race condition in TP agent"


TSAN was (rightfully!) detecting and complaining about a race due to the fact that upon init the TP agent exchanges the device maps between nodes using RPC requests (and by doing so it accesses the device maps) and then sets the reverse device maps (thus possibly modifying the set of devices). This resulted in a data race, i.e., simultaneously reading and writing the set of devices without synchronizing.

One solution is to add a mutex around the devices, which works, but is "annoying". An alternative solution is to make the set of devices immutable (i.e., `const`). For that to work, we need to exchange the device maps without using RPC calls. We can do so using the process group that we need to create anyways.

Since now there's a lot more logic in Python, I've moved (and restructured) all safety checks over there, and removed them from C++.

Differential Revision: [D28603754](https://our.internmc.facebook.com/intern/diff/D28603754/)

[ghstack-poisoned]

lw committed:
Update on "Fix race condition in TP agent"


TSAN was (rightfully!) detecting and complaining about a race due to the fact that upon init the TP agent exchanges the device maps between nodes using RPC requests (and by doing so it accesses the device maps) and then sets the reverse device maps (thus possibly modifying the set of devices). This resulted in a data race, i.e., simultaneously reading and writing the set of devices without synchronizing.

One solution is to add a mutex around the devices, which works, but is "annoying". An alternative solution is to make the set of devices immutable (i.e., `const`). For that to work, we need to exchange the device maps without using RPC calls. We can do so using the process group that we need to create anyways.

Since now there's a lot more logic in Python, I've moved (and restructured) all safety checks over there, and removed them from C++.

Differential Revision: [D28603754](https://our.internmc.facebook.com/intern/diff/D28603754/)

[ghstack-poisoned]

lw committed:
Update on "Fix race condition in TP agent"


TSAN was (rightfully!) detecting and complaining about a race due to the fact that upon init the TP agent exchanges the device maps between nodes using RPC requests (and by doing so it accesses the device maps) and then sets the reverse device maps (thus possibly modifying the set of devices). This resulted in a data race, i.e., simultaneously reading and writing the set of devices without synchronizing.

One solution is to add a mutex around the devices, which works, but is "annoying". An alternative solution is to make the set of devices immutable (i.e., `const`). For that to work, we need to exchange the device maps without using RPC calls. We can do so using the process group that we need to create anyways.

Since now there's a lot more logic in Python, I've moved (and restructured) all safety checks over there, and removed them from C++.

Differential Revision: [D28603754](https://our.internmc.facebook.com/intern/diff/D28603754/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -190,81 +190,111 @@ def _tensorpipe_construct_rpc_backend_options_handler(
+                f"device count = {all_device_counts[node]}"
+            )
+
+    for source_node in all_names:
```
cool, this is very comprehensive.

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -190,81 +190,111 @@ def _tensorpipe_construct_rpc_backend_options_handler(
+    for source_node in all_names:
+        if not set(all_device_maps[source_node].keys()).issubset(all_names):
+            raise ValueError(
+                f"Node {source_node} has invalid target node names in its device maps\n"
```
this is not covered in tests?

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -190,81 +190,111 @@ def _tensorpipe_construct_rpc_backend_options_handler(
+        devices_set = set()
+        for _, map_ in my_device_maps.items():
+            devices_set.update(map_.keys())
+        for _, map_ in reverse_device_maps.items():
```
this means if a remote worker wants to use a device on this worker, it will also appear in this workers devices, even if it is not specified? This looks good to me. Just wanna make sure this is intentional.

lw replied:
I'm just maintaining the same logic as before, hence I should be asking you if this was intentional. Though perhaps this is a good sign that the logic before was too implicit and this PR is helping readability and clarity.

lw replied:
TBH, I'd be in favor of _not_ doing this and only use the keys of the local device maps (and not count the values of the remote device maps). It will make it much easier to support elasticity. But I guess any change to this now would be BC-breaking.

mrshenli reviewed (approved):
LGTM

cc @H-Huang, this PR introduced some more usage on PG.

mrshenli reviewed (changes requested):
test failures look real.

lw committed:
Update on "Fix race condition in TP agent"


TSAN was (rightfully!) detecting and complaining about a race due to the fact that upon init the TP agent exchanges the device maps between nodes using RPC requests (and by doing so it accesses the device maps) and then sets the reverse device maps (thus possibly modifying the set of devices). This resulted in a data race, i.e., simultaneously reading and writing the set of devices without synchronizing.

One solution is to add a mutex around the devices, which works, but is "annoying". An alternative solution is to make the set of devices immutable (i.e., `const`). For that to work, we need to exchange the device maps without using RPC calls. We can do so using the process group that we need to create anyways.

Since now there's a lot more logic in Python, I've moved (and restructured) all safety checks over there, and removed them from C++.

Differential Revision: [D28603754](https://our.internmc.facebook.com/intern/diff/D28603754/)

[ghstack-poisoned]

lw committed:
Update on "Fix race condition in TP agent"


TSAN was (rightfully!) detecting and complaining about a race due to the fact that upon init the TP agent exchanges the device maps between nodes using RPC requests (and by doing so it accesses the device maps) and then sets the reverse device maps (thus possibly modifying the set of devices). This resulted in a data race, i.e., simultaneously reading and writing the set of devices without synchronizing.

One solution is to add a mutex around the devices, which works, but is "annoying". An alternative solution is to make the set of devices immutable (i.e., `const`). For that to work, we need to exchange the device maps without using RPC calls. We can do so using the process group that we need to create anyways.

Since now there's a lot more logic in Python, I've moved (and restructured) all safety checks over there, and removed them from C++.

Differential Revision: [D28603754](https://our.internmc.facebook.com/intern/diff/D28603754/)

[ghstack-poisoned]

lw commented:
> this PR introduced some more usage on PG.

Though it's possible that it will go away with per-RPC device maps? Alternatively, all of this could also be done through the store, though we need to make sure it can handle this much data.

lw committed:
Update on "Fix race condition in TP agent"


TSAN was (rightfully!) detecting and complaining about a race due to the fact that upon init the TP agent exchanges the device maps between nodes using RPC requests (and by doing so it accesses the device maps) and then sets the reverse device maps (thus possibly modifying the set of devices). This resulted in a data race, i.e., simultaneously reading and writing the set of devices without synchronizing.

One solution is to add a mutex around the devices, which works, but is "annoying". An alternative solution is to make the set of devices immutable (i.e., `const`). For that to work, we need to exchange the device maps without using RPC calls. We can do so using the process group that we need to create anyways.

Since now there's a lot more logic in Python, I've moved (and restructured) all safety checks over there, and removed them from C++.

Differential Revision: [D28603754](https://our.internmc.facebook.com/intern/diff/D28603754/)

[ghstack-poisoned]

lw committed:
Update on "Fix race condition in TP agent"


TSAN was (rightfully!) detecting and complaining about a race due to the fact that upon init the TP agent exchanges the device maps between nodes using RPC requests (and by doing so it accesses the device maps) and then sets the reverse device maps (thus possibly modifying the set of devices). This resulted in a data race, i.e., simultaneously reading and writing the set of devices without synchronizing.

One solution is to add a mutex around the devices, which works, but is "annoying". An alternative solution is to make the set of devices immutable (i.e., `const`). For that to work, we need to exchange the device maps without using RPC calls. We can do so using the process group that we need to create anyways.

Since now there's a lot more logic in Python, I've moved (and restructured) all safety checks over there, and removed them from C++.

Differential Revision: [D28603754](https://our.internmc.facebook.com/intern/diff/D28603754/)

[ghstack-poisoned]

lw committed:
Update on "Fix race condition in TP agent"


TSAN was (rightfully!) detecting and complaining about a race due to the fact that upon init the TP agent exchanges the device maps between nodes using RPC requests (and by doing so it accesses the device maps) and then sets the reverse device maps (thus possibly modifying the set of devices). This resulted in a data race, i.e., simultaneously reading and writing the set of devices without synchronizing.

One solution is to add a mutex around the devices, which works, but is "annoying". An alternative solution is to make the set of devices immutable (i.e., `const`). For that to work, we need to exchange the device maps without using RPC calls. We can do so using the process group that we need to create anyways.

Since now there's a lot more logic in Python, I've moved (and restructured) all safety checks over there, and removed them from C++.

Differential Revision: [D28603754](https://our.internmc.facebook.com/intern/diff/D28603754/)

[ghstack-poisoned]

lw committed:
Update on "Fix race condition in TP agent"


TSAN was (rightfully!) detecting and complaining about a race due to the fact that upon init the TP agent exchanges the device maps between nodes using RPC requests (and by doing so it accesses the device maps) and then sets the reverse device maps (thus possibly modifying the set of devices). This resulted in a data race, i.e., simultaneously reading and writing the set of devices without synchronizing.

One solution is to add a mutex around the devices, which works, but is "annoying". An alternative solution is to make the set of devices immutable (i.e., `const`). For that to work, we need to exchange the device maps without using RPC calls. We can do so using the process group that we need to create anyways.

Since now there's a lot more logic in Python, I've moved (and restructured) all safety checks over there, and removed them from C++.

Differential Revision: [D28603754](https://our.internmc.facebook.com/intern/diff/D28603754/)

[ghstack-poisoned]

lw committed:
Update on "Fix race condition in TP agent"


TSAN was (rightfully!) detecting and complaining about a race due to the fact that upon init the TP agent exchanges the device maps between nodes using RPC requests (and by doing so it accesses the device maps) and then sets the reverse device maps (thus possibly modifying the set of devices). This resulted in a data race, i.e., simultaneously reading and writing the set of devices without synchronizing.

One solution is to add a mutex around the devices, which works, but is "annoying". An alternative solution is to make the set of devices immutable (i.e., `const`). For that to work, we need to exchange the device maps without using RPC calls. We can do so using the process group that we need to create anyways.

Since now there's a lot more logic in Python, I've moved (and restructured) all safety checks over there, and removed them from C++.

Differential Revision: [D28603754](https://our.internmc.facebook.com/intern/diff/D28603754/)

[ghstack-poisoned]
